--- /etc/inc/authgui.inc	2025-08-23 13:22:59
+++ /etc/inc/authgui.inc	2025-08-20 16:24:48
@@ -351,6 +351,19 @@
 			                <input name="usernamefld" id="usernamefld" type="text" placeholder="Username" autocorrect="off" autocapitalize="none"/>
 			                <input name="passwordfld" id="passwordfld" type="password" placeholder="Password" />
 			                <input type="submit" name="login" value="Sign In" class="btn btn-success btn-sm" />
+
+                        <?php
+                        # Added by pfSense-pkg-saml2-auth - Show SSO login link on login page
+                        require_once("Saml2/autoload.php");
+                        $saml2_conf = new \Saml2\Config();
+
+                        # Only display the SSO login link if SAML2 auth is enabled
+                        if ($saml2_conf->enable) {
+                            echo '<link rel="stylesheet" href="/css/saml2_auth/saml2_auth.css" type="text/css">';
+                            echo '<a id="saml2_login_button" href="/saml2_auth/sso/"><p>Sign In with SSO</p></a>';
+                        }
+                        ?>
+
 		                </form>
 					</div>
 	            </div>
